Quasiquotes WIP

densh opened the pull request:
This is WIP pull request to test that history-rewritten commits are building fine before the actual pull request.

densh committed:
add missing copyTypeDef utility function

densh committed:
rename mkAnnotationCtor into mkAnnotation

densh committed:
merge flagsFromBits and FlagsAsBits into FlagsRepr

densh committed:
use NoMods and NoFlags for reification of empty values

densh committed:
move mkNew from tools.nsc.ast.TreeGen to reflect.internal.TreeGen

This is needed for quasiquotes to implement SyntacticNew combinator.

densh committed:
minor changes to TreeGen.mkTemplate

1. Make superPos parameter optional with default value NoPosition
2. use Option instead of List for local optional constructor variable

densh committed:
reduce amount of clutter on -Yquasiquote-debug

densh committed:
SI-7196 add support for refineStat splicing and extraction

xeno-by commented on src/compiler/scala/reflect/reify/utils/Extractors.scala:
```diff
@@ -175,10 +175,10 @@ trait Extractors {
             Select(Select(uref1 @ Ident(_), build1), freeTermFactory),
             _ :+
-            Apply(Select(Select(uref2 @ Ident(_), build2), flagsFromBits), List(Literal(Constant(flags: Long)))) :+
+            Apply(Select(Select(uref2 @ Ident(_), build2), flagsCtor), List(Literal(Constant(flags: Long)))) :+
```
`flagsRepr`?

densh replied:
fixed

densh committed:
refine block and applied/typeapplied splicing/matching semantics

1. blocks now match single term-level expressions to account for
   automatic block elimination. E.g.

   val q"{ ..$stats }" = q"foo"

   will match into stats = List(q"foo"). This is useful to uniformly
   deal with blocks on term level.

2. blocks in quasiquotes collapse into single expressions

3. Applied and TypeApplied now have constructors too which helps
   to unify matching and extraction in quasiquote reifier

4. TypeApplied now matches AppliedTypeTree too

5. Add Syntactic prefix to Applied and TypeApplied

densh committed:
first-class early def splicing and extraction support

densh committed:
make the postfixOps warning go away

densh committed:
add toolbox-based utility methods to quasiquotes' tests

densh committed:
refactor definition tests into separate subsuite

xeno-by commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -1458,6 +1458,8 @@ self =>
         parseOther
     }
 
+    def isTypedParam(t: Tree) = t.isInstanceOf[Typed]
```
Used anywhere else? Also why not put it in TreeInfo?

densh replied:
the whole point of extracting it out is to let quasiquotes override it.

xeno-by commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -2821,8 +2812,17 @@ self =>
 
-    def isInterface(mods: Modifiers, body: List[Tree]): Boolean =
-      mods.isTrait && (body forall treeInfo.isInterfaceMember)
+    def earlyDef(tree: Tree): Option[Tree] = tree match {
```
extractEarlyDef

densh replied:
maybe `ensureEarlyDef`? it doesn't really extract anything, just checks the shape of the tree.

xeno-by replied:
ensure implies a crash if it's not an early def

densh replied:
it does throw errors (syntax error or deprecation warning), doesn't it ?

xeno-by replied:
but then why does it return options?

densh replied:
It returns None if it encounters EmptyTree. Some on proper input and fails (by mutating error reporter and also returning None) on non-valid input that is not EmptyTree.

xeno-by replied:
why doesn't it return EmptyTree instead of None?

densh replied:
because EmptyTree won't flatten into nothing, it shall not stay in the list after https://github.com/scala/scala/pull/2898/files#L5R2800 is executed.

xeno-by replied:
it's quite non-idiomatic to have options of reflection artifacts in the compiler

densh replied:
do you know any better way to rewrite this code? it originally used lists and flatten even though option would suffice.

xeno-by replied:
granted that we can't use options, i guess it'll be either returning a list or doing `.map().filter(_ != EmptyTree)`.

retronym replied:
I don't think `Option[Tree]` is universally outlawed. If we have to represent `Some(EmptyTree)` it makes sense to me.

xeno-by commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -2842,26 +2842,21 @@ self =>
         DefDef(NoMods, nme.CONSTRUCTOR, Nil, ListOfNil, TypeTree(), Block(Nil, literalUnit))
       )
+      val parentPos = o2p(in.offset)
       val tstart0 = if (body.isEmpty && in.lastOffset < tstart) in.lastOffset else tstart
```
tstart1?

densh replied:
do you seriously propose rename `parentPos` into inpenetrable `tstart1`?

xeno-by replied:
nope, i think that it's strange to transform `tstart` to `tstart0`

xeno-by commented on src/compiler/scala/tools/nsc/transform/Constructors.scala:
```diff
@@ -174,7 +174,7 @@ abstract class Constructors extends Transform with ast.TreeDSL {
 
       val bodyOfOuterAccessor: Map[Symbol, DefDef] =
-        defBuf collect { case dd: DefDef if outerCandidatesForElision(dd.symbol) => dd.symbol -> dd } toMap
+        defBuf.collect { case dd: DefDef if outerCandidatesForElision(dd.symbol) => dd.symbol -> dd }.toMap
```
NO WAY!!!

xeno-by commented on src/compiler/scala/tools/reflect/quasiquotes/Parsers.scala:
```diff
@@ -79,7 +83,14 @@ trait Parsers { self: Quasiquotes =>
+
+      def isHole: Boolean = isIdent && isHole(in.name)
+
+      def isHole(name: Name) = holeMap.contains(name)
```
Used by anyone else?

densh replied:
Thanks for reminding me, a few more places can be converted to use `isHole(name)` instead of direct access to `holeMap`

xeno-by commented on src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala:
```diff
@@ -7,7 +7,9 @@ import scala.reflect.internal.Flags._
-  import global.build.SyntacticClassDef
+  import global.build.{SyntacticClassDef, SyntacticTraitDef, SyntacticModuleDef,
+                       SyntacticDefDef, SyntacticValDef, SyntacticVarDef, SyntacticNew,
+                       SyntacticBlock, SyntacticApplied, SyntacticTypeApplied, SyntacticFunction}
```
Sometimes I wish for Perl-like imports: `import global.build.Syntactic*` :)

xeno-by commented on src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala:
```diff
@@ -49,11 +47,45 @@ trait Reifiers { self: Quasiquotes =>
+      case Block(stats, last) =>
+        reifyBuildCall(nme.SyntacticBlock, stats :+ last)
+      case Select(id @ Ident(nme.scala_), name) if id.symbol == ScalaPackage =>
+        reifyBuildCall(nme.ScalaDot, name)
```
Hygiene strikes back )))

xeno-by replied:
Can't really see how this is a syntactic transform btw.

densh replied:
syntactic transform in reifier means something completely different, not that it's done by syntactic\* extractors

xeno-by replied:
then this needs a comment

xeno-by commented on src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala:
```diff
@@ -49,11 +47,45 @@ trait Reifiers { self: Quasiquotes =>
+      case SyntacticFunction(args, body) =>
+        reifyBuildCall(nme.SyntacticFunction, args, body)
+      case Block(stats, last) =>
+        reifyBuildCall(nme.SyntacticBlock, stats :+ last)
```
How about introducing a SyntacticBlock extractor? Otherwise this line isn't consistent with the others.

densh replied:
There is a syntactic block extractor but it has different semantics (always matches terms) and is not appropriate here.

xeno-by commented on src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala:
```diff
@@ -62,6 +94,9 @@ trait Reifiers { self: Quasiquotes =>
     }
 
+    def reifyFlags(flags: FlagSet) =
+      if (flags != 0) reifyBuildCall(nme.FlagsRepr, flags) else mirrorSelect(nme.NoFlags)
```
Why not put this in the base class?

densh replied:
It is in the base class already.

xeno-by replied:
i meant the base reifier

densh replied:
I see, maybe we should implement `NoMods` optimization for `reifyModifiers` in base class too?

xeno-by replied:
Would be great

xeno-by commented on src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala:
```diff
@@ -150,8 +189,12 @@ trait Reifiers { self: Quasiquotes =>
 
-    def ensureNoExplicitFlags(m: Modifiers, pos: Position) =
-      if ((m.flags & ExplicitFlags) != 0L) c.abort(pos, s"Can't $action modifiers together with flags, consider merging flags into modifiers")
+    val quasiExplicitFlags = ExplicitFlags & ~MUTABLE & ~TRAIT
```
Wat?

densh replied:
If you disallow to set this flags with modifiers you'll be unable to do:

`q"$mods var $name: $tpt = $rhs"`

or 

`q"$mods trait $name"`

The presence of this flags after splicing is ensured by `SyntacticTraitDef` and `SyntacticVarDef`

xeno-by replied:
The name is really unclear. I believe that this special treatment is necessary. It just needs to be documented - both in the name and in an accompanying comment.

xeno-by commented on src/compiler/scala/tools/reflect/quasiquotes/Reifiers.scala:
```diff
@@ -198,56 +232,53 @@ trait Reifiers { self: Quasiquotes =>
-          case FlagsLocation => false
-          case _ => c.abort(tree.pos, s"$flagsType or $modsType expected but ${tree.tpe} found")
+    override def reifyModifiers(m: Modifiers) =
+      if (m == NoMods) mirrorSelect(nme.NoMods)
```
Why not put this in the base class?

densh replied:
reifyModifiers are implemented differently for apply/unapply cases

and you can't really put just NoMods case

xeno-by replied:
I meant making the vanilla reifier reify NoMods as NoMods and NoFlags as NoFlags.

densh replied:
Okey, makes sense.

xeno-by commented on src/continuations/plugin/scala/tools/selectivecps/SelectiveCPSTransform.scala:
```diff
@@ -350,9 +348,9 @@ abstract class SelectiveCPSTransform extends PluginComponent with
                   val argDef = localTyper.typed(ValDef(argSym, Select(ctxRef, ctxRef.tpe.member(cpsNames.getTrivialValue))))
                   val switchExpr = localTyper.typedPos(vd.symbol.pos) {
-                    val body2 = mkBlock(bodyStms, bodyExpr).duplicate // dup before typing!
+                    val body2 = gen.mkBlock(bodyStms :+ bodyExpr).duplicate // dup before typing!
```
Yeah, this also begs for mkBlock taking 2 arguments instead.

densh replied:
what's the point of numerous overloads that doesn't save any code?

xeno-by replied:
saves writing :+ and ()

densh replied:
one of the goals of Syntactic\* combinators is to eventually supersede treegen & treebuilder. adding another overload may complicate things slightly. is it really necessary to eliminate a few characters of code in deprecated part of codebase?

xeno-by replied:
how would `mkBlock(bodyStms, bodyExpr).duplicate` look when we migrate to Syntactic?

densh replied:
Syntactic\* combinators don't do any duplication at the moment, they just combine trees and return new trees with old ones spliced-in. i.e.:

`SyntacticBlock(bodyStms :+ bodyExpr).duplicate`

xeno-by commented on src/reflect/scala/reflect/internal/BuildUtils.scala:
```diff
@@ -78,63 +70,192 @@ trait BuildUtils { self: SymbolTable =>
+    def mkTparams(tparams: List[Tree]): List[TypeDef] =
+      tparams.map {
+        case td: TypeDef => copyTypeDef(td)(mods = (td.mods | PARAM) & (~DEFERRED))
+        case other => throw new IllegalArgumentException("can't splice $other as type parameter")
```
Why not abort?

densh replied:
abort at runtime? how?

xeno-by replied:
oh shi-

xeno-by commented on src/reflect/scala/reflect/internal/BuildUtils.scala:
```diff
@@ -78,63 +70,192 @@ trait BuildUtils { self: SymbolTable =>
+          else {
+            // undo conversion from (implicit ... ) to ()(implicit ... ) when its the only parameter section
+            val vparamssRestoredImplicits = ctorVparamss match {
+              case Nil :: rest if !rest.isEmpty && !rest.head.isEmpty && rest.head.head.mods.isImplicit => rest
```
Not impressed with head.head.head.head

xeno-by commented on src/reflect/scala/reflect/internal/BuildUtils.scala:
```diff
@@ -78,63 +70,192 @@ trait BuildUtils { self: SymbolTable =>
+      def unapply(tree: Tree): Option[(Modifiers, TypeName, List[TypeDef], Modifiers, List[List[ValDef]],
+                                       List[Tree], List[Tree], ValDef, List[Tree])] = tree match {
+        case ClassDef(mods, name, tparams, UnMkTemplate(parents, selfdef, ctorMods, vparamss, earlyDefs, body))
+          if !ctorMods.isTrait && !ctorMods.hasFlag(JAVA) =>
```
This means that class defs coming from java files aren't SyntacticClassDefs?

densh replied:
Yes they are not as they don't follow the trivial construct/deconstruct invariant.

xeno-by replied:
That's too bad, because I would imagine people also wanting to handle those as well. Would it be hard to support them?

densh replied:
The point is is that you should make sure that transforms like:

``` scala
case q"$mods class $name(..$args) { ..$body }" => q"$mods class $name(..$args) { ..$body }"
```

Don't introduce any new members due to desugaring. This invariant holds for regular classes but would introduce garbage constructor for classes created manually outside of the parser. We might want to add another syntactic device for situations like that that will have it's own parsing rules and handling.

xeno-by commented on src/reflect/scala/reflect/internal/BuildUtils.scala:
```diff
@@ -178,7 +299,95 @@ trait BuildUtils { self: SymbolTable =>
+        case AppliedTypeTree(id: Ident, args @ (argtpes :+ restpe))
+          if args.length - 1 <= MaxFunctionArity && id.symbol == FunctionClass(args.length - 1) =>
+          Some((argtpes, restpe))
+        case AppliedTypeTree(Select(pack, fun), args @ (argtpes :+ restpe))
```
pkg

xeno-by commented on src/reflect/scala/reflect/internal/BuildUtils.scala:
```diff
@@ -178,7 +299,95 @@ trait BuildUtils { self: SymbolTable =>
+          if args.length - 1 <= MaxFunctionArity && id.symbol == FunctionClass(args.length - 1) =>
+          Some((argtpes, restpe))
+        case AppliedTypeTree(Select(pack, fun), args @ (argtpes :+ restpe))
+          if args.length - 1 <= MaxFunctionArity && pack.symbol == ScalaPackage && fun == FunctionClass(args.length - 1).name =>
```
Why .name here and .symbol in the previous case?

xeno-by replied:
Also this check should be deduplicated into a function

densh replied:
obviously Ident("Function10") can mean just about anything unlike Select(scala package, "Function10"), i would check for symbols everywhere but in previous discussion you for some reason said it's a bad idea as we don't know how to deal with symbols.

xeno-by commented on src/reflect/scala/reflect/internal/BuildUtils.scala:
```diff
@@ -178,7 +299,95 @@ trait BuildUtils { self: SymbolTable =>
+        case SyntacticApplied(Select(New(SyntacticTypeApplied(ident, targs)), nme.CONSTRUCTOR), argss) =>
+          Some((Nil, SyntacticApplied(SyntacticTypeApplied(ident, targs), argss) :: Nil, emptyValDef, Nil))
+        case SyntacticBlock(SyntacticClassDef(_, tpnme.ANON_CLASS_NAME, Nil, _, List(Nil), earlyDefs, parents, selfdef, body) ::
+                            Apply(Select(New(Ident(tpnme.ANON_CLASS_NAME)), nme.CONSTRUCTOR), Nil) :: Nil) =>
```
AnonClassDef?

densh replied:
doesn't warrant an extractor for a single use case especially if the code (SyntacticNew) itself is abstraction to make people never see these anonymous classes

xeno-by commented on src/reflect/scala/reflect/internal/BuildUtils.scala:
```diff
@@ -178,7 +299,95 @@ trait BuildUtils { self: SymbolTable =>
+        case ValDef(mods, name, tpt, rhs) if mods.hasFlag(MUTABLE) => Some((mods, name, tpt, rhs))
+        case _ => None
+      }
+    }
```
Copy/paste

densh committed:
unify handling of references to scala members for functions and tuples

densh committed:
reify ScalaPackage symbol with the help of ScalaDot

densh committed:
add support for function type splicing and extraction

densh committed:
SI-7723 better support for deconstruction of new expressions

densh committed:
rename TupleN and TupleTypeN into SyntacticTuple and SyntacticTupleType

densh committed:
better support for ValDefs, VarDefs and DefDefs

densh committed:
SI-7803 support for matching of anonymous function literals

densh committed:
streamline implementation of annotation splicing

Syntax spec mislead me to believe that annotation can't have type
parameters or multiple argument lists... I guess the lesson here is
don't trust the spec.
